src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Managed.cs

```diff
@@ -175,6 +176,26 @@ public async Task ConnectAsync(Uri uri, CancellationToken cancellationToken, Cli
                     }
                 }
 
+                // Because deflate options are negotiated we need a new object
+                WebSocketDeflateOptions? deflateOptions = null;
+
+                if (options.DeflateOptions is not null && response.Headers.TryGetValues(HttpKnownHeaderNames.SecWebSocketExtensions, out IEnumerable<string>? extensions))
+                {
+                    foreach (ReadOnlySpan<char> extension in extensions)
+                    {
+                        if (extension.TrimStart().StartsWith(ClientWebSocketDeflateConstants.Extension))
+                        {
+                            deflateOptions = ParseDeflateOptions(extension, options.DeflateOptions);
+                            break;
+                        }
+                    }
+                }
+
+                // Store the negotiated deflate options in the original options, because
+                // otherwise there is now way of clients to actually check whether we are using
+                // per message deflate or not.
+                options.DeflateOptions = deflateOptions;
```

Comment:
Do we have prior-art for this kind of behavior?
Is it important for users to know what was negotiated?

Reply:
I am not aware of, but it seemed reasonable to me because the `ClientWebSocketOptions` is a property inside the `ClientWebSocket` and it seemed wrong to me to have deflate options there which don't represent the actual state.

Also if we don't it, I can't really test the negotiation of the deflate, without exposing/making other properties.

Reply:
Can these options be shared across multiple WebSocket instances?

Reply:
If you're asking about `ClientWebSocketOptions` then the answer is no, because they are created in place by the `ClientWebSocket`. If you mean `WebSocketDeflateOptions` I see no reason why then can't be shared across instances.

Reply:
A good comparison here, I think, is subprotocol handling. You can define what subprotocols you want to support on the ClientWebSocketOptions class. but after this is negotiated, the chosen subprotocol is exposed via ClientWebSocket.SubProtocol. We should do something similar here.

Reply:
@scalablecory Correct, this isn't the case. The options being mutated here is the instance that belongs solely to ClientWebSocket. I can also do what @geoffkizer suggests, but this requires new API to be introduced.

Reply:
I revert my position here. Here is a situation where things could blow up unexpectedly:
```c#
var client = new ClientWebSocket();
client.Options.DangerousDeflateOptions = new();

await client.ConnectAsync();

// Because we have nullable annotations, if we try to use the deflate options on the next line, 
// the compilator will be happy, because it sees the instance has been initialized beforehand. 
// However if the server we're connecting to doesn't support deflate, my current implementation 
// would reset the instance to null.
var clientWindowBits = client.Options.DangerousDeflateOptions.ClientMaxWindowBits;
```

The only way in my opinion to do this safely and expose the negotiated options to users is to do what @geoffkizer suggested and create a property in WebSocket `DangerousDeflateOptions`. It's not really necessary in regards to the implementation, but using ClientWebSocket, users will not be able to tell if compression is being used or not.

@scalablecory What do you think?

Reply:
Do we have a real world need for client knowing if their ask for compression was honored? Can we add it later if a need arises?

Reply:
Something like

```csharp
WebSocketDeflateOptions? NegotiatedDeflateOptions { get; }
```

Reply:
@benaadams exactly like this. 

@scalablecory without it I don't know how I am supposed to write tests that make sure client and server negotiate the options correctly? I guess I can create an internal property and use [InternalsVisibleTo] but that looks very dirty to me.

Reply:
Although I am not sure we need the Negotiated prefix, because it is the same with the SubProtocol property, the client can send multiple options and the server can choose.